topdown/builtins.go

```diff
@@ -233,6 +241,60 @@ func functionalWrapper1Out3(name ast.String, fn FunctionalBuiltin1Out3) BuiltinF
 	}
 }
 
+func userFunctionWrapper(name ast.String, fns []*ast.Func) BuiltinFunc {
+	return func(t *Topdown, expr *ast.Expr, iter Iterator) error {
+		operands := expr.Terms.([]*ast.Term)[1:]
+		resolved, err := resolveN(t, name, operands, len(operands)-1)
+		if err != nil {
+			return err
+		}
+
+		var rTerms ast.Array
+		for _, r := range resolved {
+			rTerms = append(rTerms, ast.NewTerm(r))
+		}
+
+		var result *ast.Term
+		for _, fn := range fns {
+			child := t.Child(fn.Body)
+			child.traceEnter(fn)
+
+			evalErr := &Error{
+				Code:    ConflictErr,
+				Message: fmt.Sprintf("function %s produces conflicting outputs", name),
+			}
+
+			arr := ast.Array(fn.Head.Args)
+			undo, err := evalEqUnify(child, rTerms, arr, nil, true, func(child *Topdown) error {
+				return eval(child, func(child *Topdown) error {
+					next := PlugTerm(fn.Head.Output, child.Binding)
+					if result != nil && !result.Equal(next) {
+						return evalErr
+					}
+					result = next
+					child.traceExit(fn)
+					return nil
+				})
+			})
+			defer child.Unbind(undo)
+
+			if err != nil && err == evalErr {
```

Comment:
If there was a unification error, we need to keep searching for a function that matches the given arguments.